File: go/022.go
package main

/*
括号生成 数字n代表生成括号的对数，请你设计一个函数，用于能够生成所有可能的并且有效的括号组合。
思路：n个左括号和右括号的全排列，然后看是否符合括号匹配。
全排列 + 是否匹配
双递归 递归确定左括号然后递归确定右括号l r，优先确定左括号，然后当有左括号放的前提下可放右括号
左括号 l > 0
右括号 l < r
*/

func generateParenthesis(n int) []string {
	res := make([]string, 0)
	if n == 0 {
		return res
	}
	find(n, n, "", &res)
	return res
}
func find(l, r int, s string, res *[]string) {
	if l == 0 && r == 0 {
		*res = append(*res, s)
		return
	}
	if l > 0 {
		find(l-1, r, s+"(", res)
	}
	if l < r {
		find(l, r-1, s+")", res)
	}
}

File: c++/022.cpp
class Solution {
/*
[简单易懂的动态规划](https://leetcode-cn.com/problems/generate-parentheses/solution/zui-jian-dan-yi-dong-de-dong-tai-gui-hua-bu-lun-da/)
*/
public:
    vector<string> generateParenthesis(int n) {
         vector<string> res;
        string str;
        stack<string> s;
        stack<int> l, r;
        int L=1, R=0;
        char c1='(', c2=')';
        str+=c1;
        s.push(str);
        l.push(L);
        r.push(R);
        while(!s.empty()){
            str = s.top();
            L = l.top();
            R = r.top();
            s.pop();
            l.pop();
            r.pop();
            if((L==n)&&(R==n)){
                res.push_back(str);
                continue;
            }
            if(L < n ){
                str = str + c1;
                L++;
                s.push(str);
                l.push(L);
                r.push(R);
                str.erase(str.length()-1);
                L--;
            }
            if(R < n && L > R){
                str = str + c2;
                R++;
                s.push(str);
                r.push(R);
                l.push(L);
            }
        }
        return res;
    }
};